### ISSUE TITLE
Syntax error in text mermaid version 11.4.1

### ISSUE THREAD

**Issue by chenyang-shanghai** [ahmedkhaleel2004/gitdiagram#29]
In the step 3/3, "Syntax error in text mermaid version 11.4.1" occurs frequently.
Maybe we should state In the prompt that the LLM is required to strictly adhere to the Mermaid syntax.

**Comment by rachidje** [comment]
Same here with repo coding in Python (fastapi)

**Comment by Chrys4lis01** [comment]
Same here

**Comment by pr0pz** [comment]
Same here with one of my repos

**Comment by Voidless7125** [comment]
Same here with my repos that contain C++. It's like they are trying to build it?

**Comment by stanoswald** [comment]
Run into same problem here

**Comment by danielsole** [comment]
Same here with any repo

**Comment by RealmiRoblox** [comment]
Same problem

**Comment by ahmedkhaleel2004** [comment]
This error simply means the LLM generated incorrect Mermaid.js diagram code.

I've tried addressing this issue in numerous ways but ultimately I find that there is no reliable way. It is mostly a limitation of the LLM. If there is a way to validate Mermaid.js code, that would help but as of now I am not sure how.

> Maybe we should state In the prompt that the LLM is required to strictly adhere to the Mermaid syntax.

I already do this, for the errors that I've specifically found. You can see it in the `prompts.py` under the third prompt (that generates the Mermaid.js code), but of course this is not guaranteed and new errors always show up.

**Comment by Voidless7125** [comment]
> This error simply means the LLM generated incorrect Mermaid.js diagram code.
> 
> I've tried addressing this issue in numerous ways but ultimately I find that there is no reliable way. It is mostly a limitation of the LLM. If there is a way to validate Mermaid.js code, that would help but as of now I am not sure how.
> 
> > Maybe we should state In the prompt that the LLM is required to strictly adhere to the Mermaid syntax.
> 
> I already do this, for the errors that I've specifically found. You can see it in the `prompts.py` under the third prompt (that generates the Mermaid.js code), but of course this is not guaranteed and new errors always show up.
> 
> 
> 

Thanks for the reply. I've used this before, but it was a few months ago and it worked fine. I fully understand why it might take time (or may never get fixed) but thank you!

**Comment by gitnasr** [comment]
That's an LLM issue; most LLMs still can't generate an error-free Mermaid Diagram.

I suggest, as a fallback, that you return to the user the Incorrect Code, The user could tune it or give it back to another LLM to tune and fix it. That will be beneficial 

@ahmedkhaleel2004

**Comment by gitnasr** [comment]
> That's an LLM issue; most LLMs still can't generate an error-free Mermaid Diagram.
> 
> I suggest, as a fallback, that you return to the user the Incorrect Code, The user could tune it or give it back to another LLM to tune and fix it. That will be beneficial
> 
> [@ahmedkhaleel2004](https://github.com/ahmedkhaleel2004)

Opened #34, @ahmedkhaleel2004. Can you please check it out?

**Comment by h8ngryDev4Hire** [comment]
Is there a way for the mermaid.js code to be ran through a separate bash process where code valdiation is determined by the exit code and the llm can self-correct based on the error stack? (W/out having taken a look at the current repo structure, Im assuming that this is a possibility)

**Comment by nithinworks** [comment]
Hey may I know what LLM you r using? Mistral AI codestral one generates error free mermaids most of the time. I used the same model in one of my app code to flow charts.

**Comment by ahmedkhaleel2004** [comment]
> what LLM you r using

openai o3-mini

**Comment by jonathanbertholet** [comment]
> > what LLM you r using
> 
> openai o3-mini

It might be worth giving Gemini a shot with mermaid documentation in cache. The large context window usually solves these issues.

**Comment by hizmarck** [comment]
Also here

![Image](https://github.com/user-attachments/assets/af866d9b-683e-45c7-9548-08201058fd45)

**Comment by Toastaspiring** [comment]
I tried to get as much info as I could on this issue by looking at the doc and testing on my side (https://mermaid.js.org/intro/syntax-reference.html)

- small repository seems to be ok
- medium repo has trouble if files are long, it get lost easily
- large repo has 100% chance of not displaying

no matter the custom instruction, we can't influence the 3rd step before the rendering... the custom instruction only seems to affect the 1st step

**Comment by 18121A05L2** [comment]
@ahmedkhaleel2004 , i was also facing this issue

**Comment by bdfsaraiva** [comment]
Same here!

**Comment by Draudnaut** [comment]
Same Issue. Do we have a solution yet?

**Comment by msurguy** [comment]
Same issues for my repos

**Comment by iambumblehead** [comment]
same :)

**Comment by puravparab** [comment]
Adding a custom instruction like "Fix syntax error" fixes the mermaid errors for me

**Comment by HookDing** [comment]
+1

This repository : https://github.com/HeroTransitions/Hero

In step 3 also throw this Exception

**Comment by JanisZhang** [comment]
Same，https://gitdiagram.com/JanisZhang/PR-conflicts-alert

**Comment by mbelveder** [comment]
+1

**Comment by cr2007** [comment]
+1

**Comment by GloomyDay** [comment]
Same for https://github.com/immobiliare/backstage-plugin-gitlab 

![Image](https://github.com/user-attachments/assets/58fefc8d-4a12-46c9-8849-8178a689a5f7)

**Comment by jcubic** [comment]
Same error:

![Image](https://github.com/user-attachments/assets/c1e57664-c8ef-4f96-9419-7dfd433077a0)

**Comment by ansar062** [comment]
Same issue

**Comment by Voidless7125** [comment]
Let not create a lot of messages saying same here, we know. If you want to add, use the thumbs up to indicate it.

I know I already mentioned this already, but @ahmedkhaleel2004, around Dec 25 2024, this worked just fine. I am not sure if you still used a LLM on it, but I never had a issue with it, and used it on 15+ repos.

**Comment by andrew76214** [comment]
Same error

**Comment by yufansong** [comment]
Hey guys, see my reply here: https://github.com/ahmedkhaleel2004/gitdiagram/issues/4#issuecomment-2772172320

In my opinion, the error result from the lack of modal ability. The LLM generated mermaid js is not correct. You have the get what the generated `mermaid.js` is and correct it. But the author didn't provide the `mermard.js` to us. Actually the frontend have shown the generated `mermaid` content in stream. But I think it is hard for people to capture and save it.

My temprory solution is, in my branch here: https://github.com/yufansong/gitdiagram/blob/feat/local-service/backend/app/routers/generate.py#L36 I directly save the generated `mermaid` context as a file. When I deploy it locally, if I find the `syntax` error, I just get this file and use LLM correct it again to get output. In order to get the file, you need to run my hacky branch and follow the instruction: https://github.com/yufansong/gitdiagram/tree/feat/local-service?tab=readme-ov-file#how-to-run-it-on-local-direcotry.

To improve it, we can just add one extra round LLM interaction. If program find mermaid rendering failed, just put error message and original `mermaid.js` into LLM, and directly get the correct answer from LLM. But I am little busy and don't familiar with frontend things. I don't think I've had much time for that recently. Hope anyone in open source community can help to improve it.

**Comment by WilliamXuanYu** [comment]
Same problem, greatly affects the experience. The number of failures far exceeds the number of results. Especially for mature (complex) projects, the minor errors caused by LLM will continue to accumulate. Continuous interaction with LLM and self-correction may be useful, but this requires a lot of tokens.

**Comment by wen495713** [comment]
can we change the model or use openrouter api?

**Comment by rite2hhh** [comment]
Have you tried specifying the mermaid version you are using in the prompt?
https://github.com/ahmedkhaleel2004/gitdiagram/blob/main/backend/app/prompts.py doesn't have any mention of it.

Since you have specified [mermaid": "^11.4.1](https://github.com/ahmedkhaleel2004/gitdiagram/blob/main/package.json#L37) in package.json, maybe it's worth restricting the model to generate the syntax for that version.

**Comment by yyz945947732** [comment]
same Error

![Image](https://github.com/user-attachments/assets/621b02b9-2a7b-4f5b-8ef2-5921904c3f84)

**Comment by liuxiriver** [comment]
aha, Same error

**Comment by nanakura** [comment]
Same error

**Comment by Vitus213** [comment]
Same here

**Comment by 18651619390** [comment]
Same here

**Comment by BiFangKNT** [comment]
Same here

**Comment by jaepil-choi** [comment]
s h

**Comment by liuxxxu** [comment]
I encountered this issue as well, but I extracted and concatenated the chunk values from the "status" field with a value of "diagram_chunk" in the stream request response body. Then I pasted it into the Mermaid editor, and it generated correctly without any syntax errors. So, I suggest that instead of showing an error on the page when this issue occurs, the system should output the corresponding Mermaid code. This way, users can fix it themselves.

**Comment by gitnasr** [comment]
> I encountered this issue as well, but I extracted and concatenated the chunk values from the "status" field with a value of "diagram_chunk" in the stream request response body. Then I pasted it into the Mermaid editor, and it generated correctly without any syntax errors. So, I suggest that instead of showing an error on the page when this issue occurs, the system should output the corresponding Mermaid code. This way, users can fix it themselves.

Already, there's a fork for the Marmaid Code when error.

**Comment by FirepadCN** [comment]
S H ,https://github.com/ellanjiang/gameframework

**Comment by rakshithvk19** [comment]
A simple fix that worked for me:
Added the following prompt in the customize diagram section --> Ignore the syntax issue from mermaid version 11.4.1 and regenerate the remainder of the diagram. 
This would at least show me something which would help me understand the codebase on a high-level even though few of the components might be missing.

**Comment by wodin** [comment]
@ahmedkhaleel2004 maybe it would be a good idea to display some sort of "This is a known problem" message when this error is caught?
Otherwise people are going to keep creating duplicate issues (or making "me too" comments here)

**Comment by birbilis** [comment]
The commercial mermaid version has an AI fixer btw. So it could be possible to pass the generated mermaid diagram a 2nd time through an LLM and tell it to fix any errors. Ideally if you could pass first the output through a mermaid validator (if any out there) and give the LLM the result from the validation to do fixes. Then loop till no more fixes or a max number of loop steps reached. Ideally would use an LLM trained on correct mermaid diagrams (possibly generated via some diagram to mermaid export from some repo of existing diagrams, or automating visio or similar to make diagrams, export to mermaid, then train or finetune some existing LLM)

**Comment by mrmartuz** [comment]
To visualize my private repo i followed this step: 
- F12 console,
- select Network tab
- select stream on the bottom left
- open the response
- waited the last line (for me was of 13544 characters)
- copy in your prefered a AI and ask to resolve any issues
- copy in [https://mermaid.live/](url)

![Image](https://github.com/user-attachments/assets/ad648220-246e-4e78-97d8-d4342b2f3538)

my errors were:

Modified Problematic Label:
Original: app/api/trpc/[trpc]
New: app/api/trpc/...

Reversed Arrow Direction for J and G:
Original: J[(PostgreSQL<br>prisma/)]:::database <--|Prisma ORM| G
New: J[(PostgreSQL<br>prisma/)]:::database -->|Prisma ORM| G

Moved Labels to Inline Arrows:
Original: linkStyle 0 stroke:#333,stroke-width:2px,label:"fetchList"
New: Removed label from linkStyle and added it to the arrow: A -->|fetchList| B
Reason: The linkStyle label attribute seems to be causing persistent parser errors, possibly due to a bug or strict parsing in your Mermaid renderer. Moving labels to inline arrow annotations (e.g., -->|label|) is a standard, well-supported Mermaid syntax that avoids linkStyle issues while preserving the diagram’s clarity.


hope this help!
probably some parts of my app are still missing
gitdiagram is awesome

**Comment by Eliaxs1900** [comment]
> This error simply means the LLM generated incorrect Mermaid.js diagram code.
> 
> I've tried addressing this issue in numerous ways but ultimately I find that there is no reliable way. It is mostly a limitation of the LLM. If there is a way to validate Mermaid.js code, that would help but as of now I am not sure how.
> 
> > Maybe we should state In the prompt that the LLM is required to strictly adhere to the Mermaid syntax.
> 
> I already do this, for the errors that I've specifically found. You can see it in the `prompts.py` under the third prompt (that generates the Mermaid.js code), but of course this is not guaranteed and new errors always show up.

It might be a good idea to embed all the documentation beforehand, prior to generating the Mermaid chart, so hat it has it 'fresh' in memory -- or alternatively, to have a model pre-trained with fine tunning solely and exclusively for that purpose

**Comment by TheKitty** [comment]
An additional example 

<img width="767" alt="Image" src="https://github.com/user-attachments/assets/f5d0003d-05ca-4782-8eed-55deea337df7" />

**Comment by 0xUINTBEEF** [comment]
same here. 

![Image](https://github.com/user-attachments/assets/4797795f-e302-4878-b458-35396539ef2b)

**Comment by deadmau5v** [comment]
I tried many repositories, almost all of them have this bug

**Comment by macklinhrw** [comment]
same with https://github.com/excalidraw/excalidraw

**Comment by nayananandofficial** [comment]
yeah even with this repo: https://github.com/ahmedkhaleel2004/gitdiagram

**Comment by nastaran00** [comment]
It doesn’t work for any Git repo! That’s a major contradiction with your slogan.

**Comment by SeedyROM** [comment]
I copied the mermaid.js syntax and brought it into [mermain.live](https://mermaid.live): It looks like it's adding quotes around the click events:

<img width="615" alt="Image" src="https://github.com/user-attachments/assets/eeee5655-0acd-4d05-9807-9b95c90c4fdc" />

So it's mostly just a prompting issue that be easily fixed. But that's a guess? Maybe I'm wrong?

**Comment by mz-dz** [comment]
@claude

**Comment by mz-dz** [comment]
I think the problem is with the tokens of the AI model

The problem might be from the max tokens because if the max tokens were low, the AI wouldn't be able to complete typing because it would end before finishing

@ahmedkhaleel2004

**Comment by muke78** [comment]
Testing a little the generation of the mermaid.js diagram and copying it, I got this peculiar error
<img width="1070" height="585" alt="Image" src="https://github.com/user-attachments/assets/ab57c79a-684e-42c2-a26a-58d82b2fa577" />

<img width="1070" height="585" alt="Image" src="https://github.com/user-attachments/assets/9ba71e62-4b5a-489a-9351-e2cd45e3913a" />

The error is that the style property is not recognized. It's being recognized more as CSS than mermaid.

Using Claude's AI, he told me what problem he was having, and that was the problem from what I see when using OpenAI O3-Mini. @ahmedkhaleel2004  It is deficient in correcting certain syntax that does not give the form, but is correct. Perhaps with a slightly more advanced Gemini or Claude model it could work better.

**Comment by birbilis** [comment]
maybe add some "linting" step to clean up / fix known errors?
